resources/assets/lib/osu-common.ts

```diff
@@ -0,0 +1,372 @@
+        return `${array.slice(0, -1).join(osu.trans(`${key}.words_connector`))}${osu.trans(`${key}.last_word_connector`)}${String(array[array.length - 1])}`;
+    }
+  },
+
+  transChoice: (key: string, count: number, replacements: Record<string, unknown> = {}, locale?: string): string => {
+    locale ??= currentLocale;
+    const isFallbackLocale = locale === fallbackLocale;
+
+    if (!isFallbackLocale && !osu.transExists(key, locale)) {
+      return osu.transChoice(key, count, replacements, fallbackLocale);
+    }
+
+    replacements.count_delimited = osu.formatNumber(count, null, null, locale);
+    const translated = Lang.choice(key, count, replacements, locale);
+
+    if (!isFallbackLocale && translated == null) {
+      // added by Lang.choice
+      delete replacements.count;
+
+      return osu.transChoice(key, count, replacements, fallbackLocale);
+    }
+
+    return translated;
+  },
+
+  // Handles case where crowdin fills in untranslated key with empty string.
+  transExists: (key: string, locale?: string) => {
+    const translated = Lang.get(key, null, locale);
+
+    return osu.present(translated) && translated !== key;
+  },
+
+  updateQueryString: (url: string | null, params: { [key: string]: string | null | undefined }) => {
+    const docUrl = getCurrentUrl();
+    const urlObj = new URL(url ?? docUrl.href, docUrl.origin);
+
+    // FIXME: not sure about this conversion `for own` from coffeescript
+    for (const key in params) {
+      if (Object.prototype.hasOwnProperty.call(params, key)) {
+        const value = params[key];
+
+        if (value != null) {
+          urlObj.searchParams.set(key, value);
+        } else {
+          urlObj.searchParams.delete(key);
+        }
+      }
+    }
+
+    return urlObj.href;
+  },
+
+  // Wrapping the string with quotes and escaping the used quotes inside
+  // is sufficient. Use double quote as it's easy to figure out with
+  // encodeURI (it doesn't escape single quote).
+  urlPresence: (url?: string | null) => osu.present(url) ? `url("${String(url).replace(/"/g, '%22')}")` : undefined,
+
+  urlRegex: /(https?:\/\/((?:(?:[a-z0-9]\.|[a-z0-9][a-z0-9-]*[a-z0-9]\.)*[a-z][a-z0-9-]*[a-z0-9](?::\d+)?)(?:(?:(?:\/+(?:[a-z0-9$_\.\+!\*',;:@&=-]|%[0-9a-f]{2})*)*(?:\?(?:[a-z0-9$_\.\+!\*',;:@&=-]|%[0-9a-f]{2})*)?)?(?:#(?:[a-z0-9$_\.\+!\*',;:@&=/?-]|%[0-9a-f]{2})*)?)?(?:[^\.,:\s])))/ig,
+
+  uuid: () => Turbolinks.uuid(), // no point rolling our own
+
+  xhrErrorMessage: (xhr: JQuery.jqXHR) => {
+    const validationMessage = xhr.responseJSON.validation_error;
+    let message: string;
+
+    if (validationMessage != null) {
+      let allErrors: string[] = [];
+
+      // FIXME: not sure about this conversion `for own` from coffeescript
+      for (const field in validationMessage) {
+        if (Object.prototype.hasOwnProperty.call(validationMessage, field)) {
+          const errors = validationMessage[field];
+          allErrors = allErrors.concat(errors);
+        }
+      }
```

Comment:
Same here about `for own`.

Original coffeescript:

```coffeescript
      for own _field, errors of validationMessage
        allErrors = allErrors.concat(errors)
```

Reply:
`validation_error` should just be a plain object so you can use `for (const [a, b] of Object.entries(obj))` or `Object.values` if only the values are needed.